lib/Sema/CSDiagnostics.h

```diff
@@ -2102,6 +2084,88 @@ class UnableToInferKeyPathRootFailure final : public FailureDiagnostic {
   bool diagnoseAsError() override;
 };
 
+class AbstractRawRepresentableFailure : public FailureDiagnostic {
+protected:
+  Type RawReprType;
+  Type ExpectedType;
+
+  AbstractRawRepresentableFailure(const Solution &solution, Type rawReprType,
+                                  Type expectedType, ConstraintLocator *locator)
+      : FailureDiagnostic(solution, locator),
+        RawReprType(resolveType(rawReprType)),
+        ExpectedType(resolveType(expectedType)) {}
+
+public:
+  virtual Type getFromType() const = 0;
+  virtual Type getToType() const = 0;
+
+  bool diagnoseAsError() override;
+  bool diagnoseAsNote() override;
+
+protected:
+  Optional<Diag<Type, Type>> getDiagnostic() const;
+
+  virtual void fixIt(InFlightDiagnostic &diagnostic) const = 0;
+};
+
+/// Diagnose an attempt to initialize raw representable type or convert to it
+/// a value of some other type that matches its `RawValue` type.
+///
+/// ```swift
+/// enum E : Int {
+///   case a, b, c
+/// }
+///
+/// let _: E = 0
+/// ```
+///
+/// `0` has to be wrapped into `E(rawValue: 0)` and either defaulted via `??` or
+/// force unwrapped to constitute a valid binding.
+class MissingRawRepresentativeInitFailure final
+    : public AbstractRawRepresentableFailure {
+public:
+  MissingRawRepresentativeInitFailure(const Solution &solution,
+                                      Type rawReprType, Type expectedType,
+                                      ConstraintLocator *locator)
+      : AbstractRawRepresentableFailure(solution, rawReprType, expectedType,
+                                        locator) {}
+
+  Type getFromType() const override { return ExpectedType; }
+  Type getToType() const override { return RawReprType; }
+
+protected:
+  void fixIt(InFlightDiagnostic &diagnostic) const override;
+};
+
+/// Diagnose an attempt to pass raw representable type where its raw value
+/// is expected instead.
+///
+/// ```swift
+/// enum E : Int {
+///   case one = 1
+/// }
+///
+/// let _: Int = E.one
+/// ```
+///
+/// `E.one` has to use `.rawValue` to match `Int` expected by pattern binding.
+class UseOfRawRepresentableInsteadOfItsRawValueFailure final
```

Comment:
Similar to my comment about the corresponding `ConstraintFix` - this class name seems unnecessarily wordy, could we just call this `MissingRawValueFailure` or `MissingRawValueAccess`?